actix-files/src/chunked.rs

```diff
@@ -38,49 +53,52 @@ impl Stream for ChunkedReadFile {
-        if let Some(ref mut fut) = self.fut {
-            let res = match ready!(Pin::new(fut).poll(cx)) {
-                Ok(Ok((file, bytes))) => {
-                    self.fut.take();
-                    self.file = Some(file);
-
-                    self.offset += bytes.len() as u64;
-                    self.counter += bytes.len() as u64;
-
-                    Ok(bytes)
-                }
-                Ok(Err(e)) => Err(e.into()),
-                Err(_) => Err(ErrorInternalServerError("Unexpected error")),
-            };
-            return Poll::Ready(Some(res));
-        }
+        let this = self.as_mut().get_mut();
+        match this.state {
+            ChunkedReadFileState::File(ref mut file) => {
+                let size = this.size;
+                let offset = this.offset;
+                let counter = this.counter;
 
-        let size = self.size;
-        let offset = self.offset;
-        let counter = self.counter;
+                if size == counter {
+                    Poll::Ready(None)
+                } else {
+                    let mut file = file
+                        .take()
+                        .expect("ChunkedReadFile polled after completion");
 
-        if size == counter {
-            Poll::Ready(None)
-        } else {
-            let mut file = self.file.take().expect("Use after completion");
+                    let fut = spawn_blocking(move || {
+                        let max_bytes =
+                            cmp::min(size.saturating_sub(counter), 65_536) as usize;
 
-            self.fut = Some(spawn_blocking(move || {
-                let max_bytes = cmp::min(size.saturating_sub(counter), 65_536) as usize;
+                        let mut buf = Vec::with_capacity(max_bytes);
+                        file.seek(io::SeekFrom::Start(offset))?;
 
-                let mut buf = Vec::with_capacity(max_bytes);
-                file.seek(io::SeekFrom::Start(offset))?;
+                        let n_bytes = file
+                            .by_ref()
+                            .take(max_bytes as u64)
+                            .read_to_end(&mut buf)?;
 
-                let n_bytes =
-                    file.by_ref().take(max_bytes as u64).read_to_end(&mut buf)?;
+                        if n_bytes == 0 {
+                            return Err(io::ErrorKind::UnexpectedEof.into());
+                        }
 
-                if n_bytes == 0 {
-                    return Err(io::ErrorKind::UnexpectedEof.into());
+                        Ok((file, Bytes::from(buf)))
+                    });
+                    this.state = ChunkedReadFileState::Future(fut);
+                    self.poll_next(cx)
                 }
+            }
+            ChunkedReadFileState::Future(ref mut fut) => {
+                let (file, bytes) =
+                    ready!(Pin::new(fut).poll(cx)).map_err(|_| BlockingError)??;
+                this.state = ChunkedReadFileState::File(Some(file));
```

Comment:
I wonder would it decrease number of polls if spawn next chunk future right here?

Reply:
From the structure of this future. The caller is responsible to actively poll an extra round when he get Some(Bytes) from the stream to go into the file branch again. Otherwise this stream would be stuck.
And I believe downstream API would be doing this. This is an internal stream chain that I think it's better to leave this way. Using Context::waker::wake_by_ref here certainlly is a good way to self wakeup but I'm afraid it would have a liitle bit more overhead than the downstream caller handling it(or result in more polling than needed as it does not aware this change)

Reply:
I see, makes sense in that case